### PR TITLE
Added training with augmentation for VTR and TTR

### PR DESCRIPTION
**Pull request by kategimranova**

### CODE REVIEW

File: src/datasets/bert_dataset.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import torch
 from loguru import logger
 from torch.utils.data import Dataset
-from transformers import BertTokenizer
+from transformers import BertTokenizer, NllbTokenizer
```
Comment (SpirinEgor) [inline src/datasets/bert_dataset.py:4]: Replace both tokenizers with `PreTrainedTokenizer`

File: src/datasets/common.py [hunk 2]
```diff
@@ -65,7 +65,7 @@ def __init__(
             max_augmentations=max_augmentations,
         )
         if tokenizer:
-            self.tokenizer = BertTokenizer.from_pretrained(tokenizer)
+            self.tokenizer = tokenizer
         if max_seq_len:
             self.max_seq_len = max_seq_len
```
Comment (SpirinEgor) [inline src/datasets/common.py:70]:
```python
self.tokenizer = tokenizer
self.max_seq_len = max_seq_len
```
Same result

File: src/datasets/common.py [hunk 3]
```diff
@@ -54,8 +54,8 @@ def __init__(
         proba_per_text: float,
         expected_changes_per_text: int,
         max_augmentations: int,
-        tokenizer: None | str,
-        max_seq_len: None | int,
+        tokenizer: NllbTokenizer = None,
```
Comment (SpirinEgor) [inline src/datasets/common.py:57]: And here use `PreTrainedTokenizer`

File: src/datasets/translation_datasets.py [hunk 4]
```diff
@@ -1,15 +1,20 @@
...
+from transformers import NllbTokenizer
+
 
 class NLLBDataset(IterableDataset):
     def __init__(
         self,
         probas: dict,
         k: float = 0.3,
+        tokenizer: NllbTokenizer = None,
+        max_seq_len: int = None,
```
Comment (SpirinEgor) [inline src/datasets/translation_datasets.py:17]: I think it is better to keep this class without any changes and provide additional `TokenizedDataset`. So, like with `SlicesDataset` we will wrap augmented or any other dataset in it. And inside this `TokenizedDataset` we will tokenize sequence and collate into a batch

File: requirements.txt [hunk 5]
```diff
@@ -16,3 +16,4 @@ fonttools~=4.38.0
 ftfy~=6.1.1
 datasets~=2.11.0
 lpips~=0.1.4
+sentencepiece~=0.1.99
```
Comment (SpirinEgor) [inline requirements.txt:19]: I'm not sure that we need sentencepiece 🤔

File: src/main_langdetect.py [hunk 6]
```diff
@@ -82,9 +146,85 @@ def train_langdetect(args: Namespace):
     )
 
 
+def train_langdetect_ttr(args: Namespace, val_data: list = None, test_data: list = None):
```
Comment (SpirinEgor) [inline src/main_langdetect.py:149]: It seems that this function is very similar to vtr training. Let's unify them:

```python
train_dataset = NLLBDataset(...)

if args.no_augmentations:
    val_dataset = FloresDataset(lang2label=lang2label, split="dev")
    test_dataset = FloresDataset(lang2label=lang2label, split="devtest")
else:
    ...
    train_dataset = AugmentationDataset(dataset=train_dataset, ...)

if args.vtr:
    train_dataset = SlicesDataset(train_dataset, ...)
    val_dataset = SlicesDataset(val_dataset, ...)
    test_dataset = SlicesDataset(test_dataset, ...)
else:
    train_dataset = TokenizedDataset(train_dataset, ...)
    val_dataset = TokenizedDataset(val_dataset, ...)
    test_dataset = TokenizedDataset(test_dataset, ...)
```

File: src/models/tasks.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+from torch import nn
+from loguru import logger
+from torch.nn import CTCLoss
+
+from src.models.vtr.ocr import OCRHead
+from src.utils.common import PositionalEncoding, compute_ctc_loss
+from src.models.embedders.vtr import VTREmbedder
+
+
+class SequenceClassifier(nn.Module):
```
Comment (SpirinEgor) [inline src/models/tasks.py:12]: Is it the same module? Perhaps, you need to update your branch with master branch

File: scripts/generate_noisy_flores.py [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+    pairs = list(probas.keys())
+    lang2label: dict = {}
+    label2lang: dict = {}
+    label = 0
+    for pair in pairs:
+        for lang in pair.split("-"):
+            if lang not in lang2label:
+                lang2label[lang] = label
+                label2lang[label] = lang
+                label += 1
```
Comment (SpirinEgor) [inline scripts/generate_noisy_flores.py:38]: I don't think that we need this complex logic. For Flores dataset we can iterate over all samples, apply augmentations, and save in the same format. So only values for `sentence_{lang}` would change.

This is very convenient for our pipeline -- we can reuse `FloresDataset` class, but specify to load data from disc, not from HF.

File: requirements.txt (hunk 5 above)
Comment (kategimranova) [reply requirements.txt:19]: it is necessary library for NLLBtokenizer

**Pull request merged**